zaza/utilities/run_report.py

```diff
@@ -0,0 +1,196 @@
+# Copyright 2019 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for logging events for a summary report.
+
+This module contains a number of functions for logging events so that they
+can be summarised, with timings, at the end of the run.
+"""
+
+import enum
+import copy
+import logging
+import time
+import math
+import yaml
+
+run_data = None
+
+
+class EnumToStrDumper(yaml.SafeDumper):
+    """Convert Enums to str when dumping."""
+
+    def represent_data(self, data):
+        """If the value is an Enum dump the value."""
+        if isinstance(data, enum.Enum):
+            return self.represent_data(data.value)
+        return super().represent_data(data)
+
+
+class ReportKeys(enum.Enum):
+    """Keys in the Run report."""
+
+    METADATA = 'Metadata'
+    EVENTS = 'Events'
+    PCT_OF_RUNTIME = 'PCT Of Run Time'
+    ELAPSED_TIME = 'Elapsed Time'
+
+
+class EventStates(enum.Enum):
+    """States that can be registered against an event."""
+
+    START = 'Start'
+    FINISH = 'Finish'
+
+
+def _init_run_data(reset=False):
+    global run_data
+    if not run_data or reset:
+        run_data = {
+            ReportKeys.EVENTS: {},
+            ReportKeys.METADATA: {}}
+
+
+def register_event(event_name, event_state, timestamp=None):
+    """Register that event_name is at event_state.
+
+    :param event_name: Name of event
+    :type event_name: str
+    :param event_state: Name of event state (should be one of EventStates).
+    :type event_state: enum.Enum
+    :param timestamp: Seconds since epoch when event_state of event_name was
+                      reached.
+    :type timestamp: float
+    """
+    global run_data
+    _init_run_data()
```

Comment:
I think this pattern should probably be a singleton getter function.  e.g.

```python

_run_data = None

def get_run_data():
    global _run_data
    if _run_data is None:
       _run_data = init_run_data()
    return _run_data

def clear_run_data():
    global _run_data
    _run_data = None
```

then you can just do:

```python
    run_data = get_run_data()
    ...
```

in each function without needing to reference the global (as the `get_run_data()` would return a reference to it).

To do it *really* properly, a class should probably be used to enforce the singleton pattern, but it's probably not worth it here.